lib/compiler-cranelift/src/translator/code_translator.rs

```diff
@@ -552,41 +593,84 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
                     .len(),
                 "translate_call results should match the call signature"
             );
+            let func_type = environ.get_function_type(func_index).unwrap();
+            let mut results_metadata = Vec::with_capacity(func_type.results().len());
+            for result in func_type.results() {
+                results_metadata.push(if *result == WasmerType::ExternRef {
+                    ValueExtraInfo {
+                        ref_counted: true,
+                        ..Default::default()
+                    }
+                } else {
+                    Default::default()
+                });
+            }
+            let (args, args_metadata) = state.peekn(num_args);
+            // work around borrow checker
+            let args = args.to_vec();
+            let args_metadata = args_metadata.to_vec();
             state.popn(num_args);
-            state.pushn(inst_results);
+            state.pushn(inst_results, &results_metadata);
+
+            // decrement the ref count of all ref-counted values as we pop them off the stack
+            for arg in args
+                .iter()
+                .zip(args_metadata.iter())
+                .filter(|(_, metadata)| metadata.ref_counted)
+                .map(|(arg, _)| arg)
+            {
+                // TODO: figure out if we can avoid this
+                environ.translate_externref_dec(builder.cursor(), *arg)?;
+            }
```

Comment:
The callee has to be responsible for dec-ref'ing all its locals. Consider a) caller does an inc-ref as part of a call b) callee does a local.set on top of the value causing a dec-ref c) we have no idea whether to dec-ref again or not. So, the callee has to do it on exit, and we don't do this loop after calls.

Reply:
Oh good point, I was confusing arguments with the stack. We already do inc/dec logic for locals on function enter/exit. The extra inc/dec in call is a no-op. I'll remove both 👍